source/mir/random/tinflex/package.d

```diff
@@ -0,0 +1,207 @@
+first two derivatives and a partitioning into intervals with at most one inflection
+point.
+
+Params:
+    f0 = probability density function of the distribution
+    f1 = first derivative of f0
+    f2 = second derivative of f0
+    c = T_c family
+    points = non-overlapping partitioning with at most one inflection point per interval
+    rho = efficiency of the Tinflex algorithm
+
+Returns:
+    Tinflex Generator.
+*/
+Tinflex!(F0, S) tinflex(F0, F1, F2, S)
+               (in F0 f0, in F1 f1, in F2 f2,
+                S c, S[] points, S rho = 1.1)
+{
+    import mir.random.tinflex.internal.calc : calcPoints;
+    // pre-calculate all the points
+    auto ips = calcPoints(f0, f1, f2, c, points, 1.1);
+    return Tinflex!(F0, S)(f0, ips, c);
+}
+
+/**
+Data body of the Tinflex algorithm.
+Can be used to sample from the distribution.
+*/
+struct Tinflex(F0, S)
+{
+    // density function
+    private const F0 _f0;
+
+    // generated partition points
+    private const IntervalPoint!S[] _ips;
+
+    // global T_c family
+    private const S c;
+
+    private this(const F0 f0, const IntervalPoint!S[] ips, const S c)
+    {
+        _f0 = f0;
+        _ips = ips;
+        this.c = c;
+    }
+
+    /// density function of the distribution
+    F0 pdf() const @property
+    {
+        return _f0;
+    }
+
+    /// density function of the distribution
+    S pdf(S x) const @property
+    {
+        return _f0(x);
+    }
+
+    /// Generated partition points
+    const(IntervalPoint!S[]) ips() @property const
+    {
+        return _ips;
+    }
+
+    /**
+    Sample n times from the distribution.
+    Params:
+        n = number of times to samples
+        rng = random number generator to use
+    Returns:
+        Array of length n with the samples
+    */
+    S opCall() const
+    {
+        import std.random : rndGen;
+        return tfSample(_f0, _ips, c, rndGen);
+    }
+
+    /// ditto
+    S opCall(RNG)(ref RNG rng) const
+        if (isUniformRNG!RNG)
+    {
+        return tfSample(_f0, _ips, c, rng);
+    }
+}
+
+/**
+Sample from the distribution.
+Params:
+    ips = calculated inflection points
+    rng = random number generator to use
+*/
+protected S tfSample(F0, S, RNG)
+          (in F0 f0, in IntervalPoint!S[] ips, in S c, ref RNG rng)
+    if (is(ReturnType!F0 == S) &&
+        isUniformRNG!RNG)
+{
+    import std.algorithm: filter, joiner, map, sum;
+
+    // TODO: filtering not needed anymore
+    auto s = ips[0..$-2].map!`a.hatA`.sum;
+    auto areas = ips[0..$-2].map!((x) => x.hatA / s);
+
+    import std.random: dice, uniform01;
+    import std.math: abs;
+    import mir.internal.math: exp;
+
+    import mir.random.tinflex.internal.transformations : inverse, antiderivative, inverseAntiderivative;
+
+    double X;
+    // acceptance-rejection sampling
+    for (;;)
+    {
+        import std.stdio;
+        writeln(areas);
+        auto j = dice(rng, areas);
```

Comment:
ok, maybe `O(1)` is something impossible, but using precomputed cumulative sums of normaized lengths, `uniform01` and binary search `O(log(n))` is possible